### PR TITLE
Iter15

### PR DESCRIPTION
**Pull request by kripsy**

### CODE REVIEW

File: internal/app/utils/utils.go [hunk 1]
```diff
@@ -3,8 +3,13 @@ package utils
 import (
 	"crypto/rand"
 	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/pkg/errors"
 )
 
+// func for generation random short URL, consist of 5 bytes
```
Comment (arttet) [inline internal/app/utils/utils.go]: Please look at comments in Go docs

File: internal/app/db/db.go [hunk 2]
```diff
@@ -219,3 +212,171 @@ func (mdb PostgresDB) CreateOrGetBatchFromStorage(ctx context.Context, batchURL
...
+	}
+
+	defer tx.Rollback()
+
+	urls := squirrel.Update("urls").
+		Set("is_deleted", true).
+		Where(squirrel.And{
+			squirrel.Eq{"short_url": shortURL},
+			squirrel.Eq{"user_id": userID},
+			squirrel.Eq{"is_deleted": nil}}).
```
Comment (arttet) [inline internal/app/db/db.go:365]: Please create a plan for this query.

File: internal/app/handlers/handlers.go [hunk 3]
```diff
@@ -267,3 +329,52 @@ func (h *APIHandler) PingDBHandler(w http.ResponseWriter, r *http.Request) {
...
+
+	str := string(body)
+
+	str = strings.ReplaceAll(str, `[`, "")
+	str = strings.ReplaceAll(str, `]`, "")
+	str = strings.ReplaceAll(str, " ", "")
+	str = strings.ReplaceAll(str, `"`, "")
+	splitstr := strings.Split(str, ",")
+	h.myLogger.Debug("Result body", zap.Any("msg", splitstr))
+	if splitstr[0] == "" {
```
Comment (arttet) [inline internal/app/handlers/handlers.go:362]: What if this array is empty?

File: internal/app/handlers/handlers.go [hunk 4]
```diff
@@ -267,3 +329,52 @@ func (h *APIHandler) PingDBHandler(w http.ResponseWriter, r *http.Request) {
...
+		h.myLogger.Debug("Bad req, splitstr[0] is empty")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+		defer cancel()
+		h.myLogger.Debug("goroutine started with urls", zap.Any("msg", splitstr))
+		err = h.repository.DeleteBatchURLFromStorage(ctx, splitstr, userID)
```
Comment (arttet) [inline internal/app/handlers/handlers.go]: It is not batch

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/app/db/db.go (hunk 2 above)
Comment (kripsy) [reply internal/app/db/db.go:365]:
```
urls=# EXPLAIN UPDATE urls SET is_deleted = True WHERE (short_url IN (
'f3d35c9c5d281ee74f2d519c62ef276e',
'88b8173eae5c008c1ffd35284f35eafe',
'cd14884494e02fc279373b8c683c82c0'
) AND user_id = 463 AND is_deleted IS NULL);
                                                                          QUERY PLAN                                                                          
--------------------------------------------------------------------------------------------------------------------------------------------------------------
 Update on urls  (cost=12.02..23.95 rows=1 width=89)
   ->  Bitmap Heap Scan on urls  (cost=12.02..23.95 rows=1 width=89)
         Recheck Cond: (short_url = ANY ('{f3d35c9c5d281ee74f2d519c62ef276e,88b8173eae5c008c1ffd35284f35eafe,cd14884494e02fc279373b8c683c82c0}'::text[]))
         Filter: ((is_deleted IS NULL) AND (user_id = 463))
         ->  Bitmap Index Scan on urls_short_url_key  (cost=0.00..12.02 rows=3 width=0)
               Index Cond: (short_url = ANY ('{f3d35c9c5d281ee74f2d519c62ef276e,88b8173eae5c008c1ffd35284f35eafe,cd14884494e02fc279373b8c683c82c0}'::text[]))
(6 rows)
```

File: internal/app/utils/utils.go (hunk 1 above)
Comment (kripsy) [reply internal/app/utils/utils.go]: 1) fix in urils.
2) add task in "TO DO"

File: internal/app/handlers/handlers.go (hunk 3 above)
Comment (kripsy) [reply internal/app/handlers/handlers.go:362]: If array empty - it returns 400.
This case i added in unit test.

File: internal/app/handlers/handlers.go (hunk 4 above)
Comment (kripsy) [reply internal/app/handlers/handlers.go]: Rename to DeleteSliceURLFromStorage

File: internal/app/db/db.go [hunk 5]
```diff
@@ -49,35 +60,21 @@ func InitDB(connString string, myLogger *zap.Logger) (*PostgresDB, error) {
...
-	ALTER TABLE IF EXISTS public.urls
-		OWNER to postgres;
-		ALTER TABLE public.urls
-ADD CONSTRAINT original_url_unq UNIQUE(original_url);
-		`
-
-	_, err := mdb.DB.ExecContext(ctx, query)
-
+func RunMigrations(ctx context.Context, connString string, myLogger *zap.Logger) error {
+	const migrationsPath = "./db/migrations"
```
Comment (arttet) [inline internal/app/db/db.go:64]: It should be a parameter.

**Review by arttet — APPROVED** [review]

**Pull request merged**